lesson16/main.js

```diff
@@ -0,0 +1,125 @@
+   loadingIcon.id = "loadingIcon";
+   loadingIcon.src = "img/loading-circle.gif";
+   body.appendChild(loadingIcon);
+};
+
+const endLoading = () => {
+   const loadingIcon = document.getElementById("loadingIcon");
+   loadingIcon.remove();
+}
+
+createTabContents();
+async function createTabContents() {
+   startLoading();
+   let responseData;
+   try {
+      responseData = await fetchData(jsonURL);
+      console.log(responseData);
+      if(responseData.length === 0) {
+         displayMessage("データはまだありません");
+         return;
+         } 
+   } catch (error) {
+      displayMessage(error);
+      console.log(error);
+   } finally {
+      endLoading();
+   }
+   createCategoryTab(responseData);
+   renderTabContainer(responseData);
+}
+
+const createCategoryTab = (values) => {
+   const fragment = document.createDocumentFragment();
+   for (let i = 0; i < values.length; i++) {
+      const categoryTabItem = createElementWithClassName("li", "category-tab__item")
+      categoryTabItem.id = `js-categoryTabItem${i}`;
+      const categoryTabText = createElementWithClassName("a", "category-tab__text");
+      categoryTabText.id = `js-categoryTabText${i}`
+      categoryTabText.textContent = values[i].category
+      fragment.appendChild(categoryTabItem).appendChild(categoryTabText);
+   }
+   categoryTab.appendChild(fragment);
+   //初期表示のタブの選択はデータとして持っているがまだ理解できていないので、現段階ではNewsのタブとしてある
+   const newsTabItem = document.getElementById("js-categoryTabItem0");
+   newsTabItem.classList.add("is-active");
+}
+
+const renderTabContainer = (values) => {
+   const tabFragment = document.createDocumentFragment();
+   const tabContainer = createElementWithClassName("div", "tab-container");
+   tabContainer.id = "js-tabContainer";
+   const articlesData = values.map((value) => value.articles);
+   for (let i = 0; i < values.length; i++) {
+      const tabContents = createElementWithClassName("div", "tab-contents");
+      tabContents.id = `js-tabContents${i}`;
+      const articlesTitles = articlesData[i].map((value) => value.title);
+      const articlesHref = articlesData[i].map((value) => value.href);
+      const tabLists = createElementWithClassName("ul", "tab-contents__lists");
+      tabLists.id = `js-tabLists${i}`;
+      const tabImage = createElementWithClassName("img", "tab-contents__image");
+      tabImage.id = `js-tabImage${i}`;
+      tabImage.src = values[i].img;
+      console.log(values);
+      console.log(articlesTitles);
+      tabContents.appendChild(tabImage);
+      for (let j = 0; j < articlesTitles.length; j++) {
+         const tabItems = createElementWithClassName("li", "lists__item");
+         const tabItemsHref = createElementWithClassName("a", "lists__item__text");
+         tabItemsHref.href = articlesHref[j];
+         tabItemsHref.textContent = articlesTitles[j];
+         tabFragment.appendChild(tabContainer).appendChild(tabContents).appendChild(tabLists).appendChild(tabItems).appendChild(tabItemsHref);
+      }
```

Comment:
IMO
I thought it would be easier to manage the article creation part if it was separated by function, and I also thought the nested loops were hard to see.

Reply:
Certainly, you're  right.
I tried to separete that articles creation part by function.
But, there may still be something wrong with them.
Please check and point out.
[8bbd5ac](https://github.com/taka0209/js_lesson/pull/18/commits/8bbd5acb6d57133ea5cd5164af932a12a9fef2c1)

Reply:
IMO
You can try using [Destructuring assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment).
```js 
const createArticlesTitles = ({articles}) => {
  const articlesTitlesFragment = document.createDocumentFragment();
  for (let i = 0; i < articles.length; i++) {
    const tabItems = createElementWithClassName("li", "lists__item");
    const tabItemsHref = createElementWithClassName("a", "lists__item__text");
    tabItemsHref.textContent = articles[i].title;
    tabItemsHref.href = articles[i].href;
    articlesTitlesFragment.appendChild(tabItems).appendChild(tabItemsHref);
  }
  return articlesTitlesFragment;
}
```

Reply:
Thank you for your review.
Q. Is it correct that we can be splitting the array as an object by specifying the elements we want in the passed array?
[f197249](https://github.com/taka0209/js_lesson/pull/18/commits/f197249a151d9ce4b2028fd6eff2cc3054f8788a)

Reply:
そうですね!
articlesオブジェクトを変数articlesに代入した という感じです。
伝わりますかね…
配列でも分割代入できますので
是非調べてみてください〜:)

Reply:
ありがとうございます！このような取り出し方ができるのを知らなかったので勉強になりました！